Snail/Snail+Combine/SnailSubscription.swift

```diff
@@ -0,0 +1,37 @@
+//  Copyright © 2021 Compass. All rights reserved.
+
+import Combine
+import Foundation
+
+@available(iOS 13.0, *)
+class SnailSubscription<Upstream: ObservableType, Downstream: Combine.Subscriber>: Combine.Subscription where Downstream.Input == Upstream.T, Downstream.Failure == Error {
+    private var disposable: Subscriber<Upstream.T>?
+    private let buffer: DemandBuffer<Downstream>
+
+    init(upstream: Upstream,
+         downstream: Downstream) {
+        buffer = DemandBuffer(subscriber: downstream)
+        disposable = upstream.subscribe(queue: nil,
+                                        onNext: { [weak self] value in
+                                            guard let self = self else { return }
+                                            _ = self.buffer.buffer(value: value)
+                                        },
+                                        onError: { [weak self] error in
+                                            guard let self = self else { return }
+                                            self.buffer.complete(completion: .failure(error))
+                                        },
+                                        onDone: { [weak self] in
+                                            guard let self = self else { return }
+                                            self.buffer.complete(completion: .finished)
+                                        })
+    }
+
+    func request(_ demand: Subscribers.Demand) {
+        _ = self.buffer.demand(demand)
+    }
+
+    func cancel() {
```

Comment:
Should this do anything with the buffer? Complete it, etc?

Reply:
This is a good question, and I'm not really sure about this right now. I'm mainly leaning on [RxCombine implementation](https://github.com/CombineCommunity/RxCombine/blob/main/Sources/Rx+Combine/RxSubscription.swift#L40) here for guidance.

I'm thinking that since demand is a concept (changing demand) that's tied with receiving values or completion, that it should really only be handled there. Since disposal could occur after a completion event, or if the subscription is removed, I don't think it would make sense to always call one of its functions here?

Reply:
That makes sense, thanks!